composite.go

```diff
@@ -0,0 +1,190 @@
+package testkit
+
+import (
+	"fmt"
+
+	"github.com/dogmatiq/testkit/assert"
+	"github.com/dogmatiq/testkit/engine/fact"
+	"github.com/dogmatiq/testkit/render"
+)
+
+// AllOf is an expectation that passes only if all of its children pass.
+func AllOf(children ...Expectation) Expectation {
+	n := len(children)
+
+	if n == 0 {
+		panic("AllOf(): at least one child expectation must be provided")
+	}
+
+	if n == 1 {
+		return children[0]
+	}
+
+	return &compositeExpectation{
+		Criteria: "all of",
+		Children: children,
+		Predicate: func(p int) (string, bool) {
+			if p == n {
+				return "", true
+			}
+
+			return fmt.Sprintf(
+				"%d of the expectations failed",
+				n-p,
+			), false
+		},
+	}
+}
+
+// AnyOf is an expectation that passes if any of its children pass.
+func AnyOf(children ...Expectation) Expectation {
+	n := len(children)
+
+	if n == 0 {
+		panic("AnyOf(): at least one child expectation must be provided")
+	}
+
+	if n == 1 {
+		return children[0]
+	}
+
+	return &compositeExpectation{
+		Criteria: "any of",
+		Children: children,
+		Predicate: func(p int) (string, bool) {
+			if p > 0 {
+				return "", true
+			}
+
+			return fmt.Sprintf(
+				"all %d of the expectations failed",
+				n,
+			), false
+		},
+	}
+}
+
+// NoneOf is an expectation that passes only if all of its children fail.
+func NoneOf(children ...Expectation) Expectation {
+	n := len(children)
+
+	if n == 0 {
+		panic("NoneOf(): at least one child expectation must be provided")
+	}
+
+	return &compositeExpectation{
+		Criteria: "none of",
+		Children: children,
+		Predicate: func(p int) (string, bool) {
+			if p == 0 {
+				return "", true
+			}
+
+			if n == 1 {
```

Comment:
```suggestion
			if p == 1 {
```

Reply:
This is correct as is. `n` is the total number of children and `p` is the number of children that passed.

This branch just changes the message that is displayed when there is only 1 child to say "the expectation passed unexpectedly" instead of "1 of the expectations passed unexpectedly".

Reply:
It might need a comment to make that clear?

Reply:
This is a closure not a func of its own. It's documented on the `Predicate` field, but I agree it could be clearer. I'll rename the vars.